src/budget/budgetmanager.cpp

```diff
@@ -957,46 +970,83 @@ int CBudgetManager::ProcessProposal(CBudgetProposal& proposal)
     return 0;
 }
 
-int CBudgetManager::ProcessProposalVote(CBudgetVote& vote, CNode* pfrom)
+bool CBudgetManager::ProcessProposalVote(CBudgetVote& vote, CNode* pfrom, int& nDos, std::string& strError)
 {
-    if (HaveSeenProposalVote(vote.GetHash())) {
-        masternodeSync.AddedBudgetItem(vote.GetHash());
-        return 0;
+    nDos = 0;
+    const uint256& voteID = vote.GetHash();
+
+    if (HaveSeenProposalVote(voteID)) {
+        masternodeSync.AddedBudgetItem(voteID);
+        return false;
     }
 
     const CTxIn& voteVin = vote.GetVin();
+
+    // See if this vote was signed with a deterministic masternode
+    auto mnList = deterministicMNManager->GetListAtChainTip();
+    auto dmn = mnList.GetMNByCollateral(voteVin.prevout);
+    if (dmn) {
+        if (!vote.CheckSignature(dmn->pdmnState->keyIDVoting)) {
+            strError = strprintf("signature invalid from dmn %s", dmn->proTxHash.ToString());
+            LogPrint(BCLog::MNBUDGET, "mvote - error: %s\n", strError);
```

Comment:
I changed this to `LogPrint`, after @furszy's suggestion, thinking about the "disk fill via logging" bug class, which is something that is currently being discussed and addressed upstream (e.g. see bitcoin#21559).

Until we have killed this bug class, the only mitigation we have is to avoid misplaced `LogPrintf` (in code paths that are easily triggered by a remote attacker).